packages/flutter/lib/src/widgets/linked_text.dart

```diff
@@ -0,0 +1,433 @@
+    return RichText(
+      text: InlineLinkedText.textLinkers(
+        style: style ?? DefaultTextStyle.of(context).style,
+        text: text,
+        textLinkers: textLinkers,
+      ),
+    );
+  }
+}
+
+/// Finds [TextRange]s in the given [String].
+typedef RangesFinder = Iterable<TextRange> Function(String text);
+
+class _TextLinkerSingle {
+  const _TextLinkerSingle({
+    required this.textRange,
+    required this.linkBuilder,
+  });
+
+  final LinkBuilder linkBuilder;
+  final TextRange textRange;
+
+  InlineSpan getSpan(String text) {
+    return linkBuilder(
+      text.substring(textRange.start, textRange.end),
+    );
+  }
+
+  /// Returns a list of [InlineSpan]s representing all of the [text].
+  ///
+  /// Ranges matched by [textLinkerSingles] are built with their respective
+  /// [LinkBuilder], and other text is represented with a simple [TextSpan].
+  static List<InlineSpan> getSpansForMany(Iterable<_TextLinkerSingle> textLinkerSingles, String text) {
+    // Sort so that overlapping ranges can be detected and ignored.
+    final List<_TextLinkerSingle> textLinkerSinglesList = textLinkerSingles
+        .toList()
+        ..sort((_TextLinkerSingle a, _TextLinkerSingle b) {
+          return a.textRange.start.compareTo(b.textRange.start);
+        });
+
+    final List<InlineSpan> spans = <InlineSpan>[];
+    int index = 0;
+    for (final _TextLinkerSingle textLinkerSingle in textLinkerSinglesList) {
+      // Ignore overlapping ranges.
+      if (index > textLinkerSingle.textRange.start) {
+        continue;
+      }
+      if (index < textLinkerSingle.textRange.start) {
+        spans.add(TextSpan(
+          text: text.substring(index, textLinkerSingle.textRange.start),
+        ));
+      }
+      spans.add(textLinkerSingle.linkBuilder(
+        text.substring(
+          textLinkerSingle.textRange.start,
+          textLinkerSingle.textRange.end,
+        ),
+      ));
+
+      index = textLinkerSingle.textRange.end;
+    }
+    if (index < text.length) {
+      spans.add(TextSpan(
+        text: text.substring(index),
+      ));
+    }
+
+    return spans;
+  }
+
+  @override
+  String toString() {
+    return '_TextLinkerSingle $textRange, $linkBuilder';
+  }
+}
+
+// TODO(justinmc): Think about which links need to go here vs. on InlineTextLinker.
+/// Specifies a way to find and style parts of a String.
+class TextLinker {
+  /// Creates an instance of [TextLinker].
+  const TextLinker({
+    // TODO(justinmc): Change "range" naming to always be "textRange"?
+    required this.rangesFinder,
+    required this.linkBuilder,
+  });
+
+  /// Builds an [InlineSpan] to display the text that it's passed.
+  final LinkBuilder linkBuilder;
+
+  /// Returns [TextRange]s that should be built with [linkBuilder].
+  final RangesFinder rangesFinder;
+
+  // TODO(justinmc): Consider revising this regexp.
+  static final RegExp _urlRegExp = RegExp(r'((http|https|ftp):\/\/)?([a-zA-Z\-]*\.)?[a-zA-Z0-9\-]*\.[a-zA-Z]*');
```

Comment:
Here's a better RegEx to use as this allows users to have port numbers, paths, query params, and fragments in the URL.

```regex
((https?|ftp):\/\/)?([a-zA-Z\-]*\.)?[a-zA-Z0-9\-]*\.[a-zA-Z]*(?::\d{1,5})?(?:\/[^\s]*)?(?:\?[^\s#]*)?(?:#[^\s]*)?
```

Reply:
@Reprevise "example." return true with your Regex

Reply:
example.

Reply:
@justinmc Your regex also fails in this case

Reply:
Thank you!  I was hoping someone would swoop in with a good regex for this.

Reply:
Oh one problem, this seems to match single periods.  Can you change it to avoid that if you have a chance?

```dart
  final RegExp regExp = RegExp(r'((https?|ftp):\/\/)?([a-zA-Z\-]*\.)?[a-zA-Z0-9\-]*\.[a-zA-Z]*(?::\d{1,5})?(?:\/[^\s]*)?(?:\?[^\s#]*)?(?:#[^\s]*)?');
  print(regExp.hasMatch('.')); // true
```

Reply:
@justinmc Hey, sorry about that. It's updated now.

Reply:
Thanks!  I confirmed that it correctly ignores `.` and `example.`.

One more problem if I can bother you again: The string "www.example.co.uk" matches only "www.example.co" and excludes the ".uk".  It works fine for the string "example.co.uk".

Also a question, should we be including ftp urls?  Nothing seems to happen for me running on Linux when I pass one through url_launcher.  Maybe I need to have an ftp client installed though?

Reply:
@justinmc No problem, here's the updated regex:

```regex
((https?|ftp):\/\/)?(([a-zA-Z\-]*\.)+[a-zA-Z]+)\.[a-zA-Z]+(?::\d{1,5})?(?:\/[^\s]*)?(?:\?[^\s#]*)?(?:#[^\s]*)?
```

In terms of FTP urls, honestly I'd skip them. I could give you a huge RegEx to allow for `username:password@example.com` with better handling of IP addresses but it's up to you. I'm 99% sure you need an FTP client installed to intercept the ftp urls btw.

Reply:
@Reprevise it seems that this regex is incorrect in many cases

Reply:
Ex: 
facebook.com
http://example.com/method?param=flutter
stackoverflow.com
https://github.com/dart-lang/dart-pad/wiki/Sharing-Guide

Reply:
@hahai96 🤦how did i miss that, i've updated it now, it matches a complete list

Reply:
This seems to work, thanks for all for all of the testing and revision!

Should I just remove the "ftp" part to ignore ftp urls?

```
(https?:\/\/)?(([a-zA-Z0-9\-]*\.)*[a-zA-Z0-9\-]+(\.[a-zA-Z]+)+)(?::\d{1,5})?(?:\/[^\s]*)?(?:\?[^\s#]*)?(?:#[^\s]*)?
```

That will do things like this:

![Screenshot 2023-05-05 at 10 38 44 AM](https://user-images.githubusercontent.com/389558/236527949-69ceede7-cd20-40ac-91b4-9d428b62ee4e.png)

Not sure if that is desired or not.

Reply:
Fine by me, though it does look weird seeing the `ftp://` not highlighted but that will be the case for other protocols too so 🤷🏻

Reply:
Maybe this one to fix the problem with the ftp not being highlighted?  I'm no regex pro and I had to repeat a section, not sure if there is a better way or not.

```
(((https?:\/\/)(([a-zA-Z0-9\-]*\.)*[a-zA-Z0-9\-]+(\.[a-zA-Z]+)+))|((?<!\/])((([a-zA-Z0-9\-]*\.)*[a-zA-Z0-9\-]+(\.[a-zA-Z]+)+))))(?::\d{1,5})?(?:\/[^\s]*)?(?:\?[^\s#]*)?(?:#[^\s]*)?
```

![Screenshot from 2023-05-08 17-21-05](https://user-images.githubusercontent.com/389558/236963678-ab5ba729-be96-453f-b84e-ad8e7ccfa3fc.png)

Reply:
Here's my attempt at fixing the "ftp://" thing:

```
(((https?:\/\/)(([a-zA-Z0-9\-]*\.)*[a-zA-Z0-9\-]+(\.[a-zA-Z]+)+))|((?<!\/|\.|[a-zA-Z0-9\-])((([a-zA-Z0-9\-]*\.)*[a-zA-Z0-9\-]+(\.[a-zA-Z]+)+))))(?::\d{1,5})?(?:\/[^\s]*)?(?:\?[^\s#]*)?(?:#[^\s]*)?
```

Now any URL with a scheme that isn't http or https won't be highlighted at all:

![Screenshot from 2023-05-09 10-09-11](https://github.com/flutter/flutter/assets/389558/5b4316ee-40ac-4891-b634-c8060bfb8531)

I had to duplicate part of the regex, so if anyone knows a better way to express that let me know.

Reply:
Hey @justinmc, here's a simpler regex that maintains the functionality:

```regex
(?<![\/\.a-zA-Z0-9-])((https?:\/\/)?(([a-zA-Z0-9-]*\.)*[a-zA-Z0-9-]+(\.[a-zA-Z]+)+))(?::\d{1,5})?(?:\/[^\s]*)?(?:\?[^\s#]*)?(?:#[^\s]*)?
```

It also matches your list exactly.

Reply:
Thank you!  Just added it and it works perfectly.  Sorry I've been working on supporting TextSpans so I missed this for a while.

Reply:
I tweaked the regex again.  Now it explicitly ignores email addresses.

![Screenshot from 2023-05-31 15-06-07](https://github.com/flutter/flutter/assets/389558/9fac396d-213c-4bc7-8d86-7b50e1a4ad91)


<details>

<summary>My list of test URLs</summary>

```dart
'www.example123.co.uk\nasdf://subdomain.example.net\nsubdomain.example.net\nftp.subdomain.example.net\nhttp://subdomain.example.net\nhttps://subdomain.example.net\nhttp://example.com/\nhttps://www.example.org/\nftp://subdomain.example.net\nexample.com\nsubdomain.example.io\nwww.example123.co.uk\nhttp://example.com:8080/\nhttps://www.example.com/path/to/resource\nhttp://www.example.com/index.php?query=test#fragment\nhttps://subdomain.example.io:8443/resource/file.html?search=query#result\n"example.com"\n\'example.com\'\n(example.com)\nsubsub.www.example.com\nhttps://subsub.www.example.com\nexample@example.com\nexample.com@example.com'
```

</details>